Model/Builders/RecurringBuilder.php

```diff
@@ -0,0 +1,255 @@
+                $this->recurringHelper->sendMail($order, 'cancel');
+                return 'success';
+            }
+            else return 'deleted';
+        }
+        return false;
+    }
+
+    public function getVaultToken()
+    {
+        $jsonContent = file_get_contents(__DIR__."/../Cron/vault_token.json");
+        return (json_decode($jsonContent, true));
+    }
+
+    public function prepareRecurringPayment()
+    {
+       // die();// This function is not properly work for now
+        $recuringData = $this->getVaultToken();
+        $oldOrder = $this->getOrderByTransaction->execute($recuringData["transactionId"]);        //  get order with all customer information by transaction id
+
+        $order = [
+            'shipping_method' =>  $oldOrder->getShippingMethod(),
+            'currency_id' => 'USD',
+            'email' => 'hello@example.com',
+            'shipping_address' => ['firstname' => 'John',
+                'lastname' => 'Green',
+                'street' => 'xxxxxx',
+                'city' => 'xxxxxxx',
+                'country_id' => 'US',
+                'region' => '1',
+                'postcode' => '85001',
+                'telephone' => '52556542',
+                'fax' => '3242322556',
+                'save_in_address_book' => 1],
+            'items' => [['product_id' => current($oldOrder->getItems())->getProductId(), 'qty' => current($oldOrder->getItems())->getQtyOrdered(), 'price' => current($oldOrder->getItems())->getPrice()]]
+        ];
+        $this->writeToFile('beforeReturn', 'd');
+        return $this->helperDataBuilder->createOrder($order);
+    }
+
+    public function prepareGingerOrder(OrderInterface $order)
+    {
+        $vaultToken = $order->getGingerpayVaultToken();
+        if (!$vaultToken)
+        {
+            return false;
+            $this->writeToFile('Bed news', 's');
+        }
+        $issuer = null;
+        $recurringType = 'recurring';
+
+        $custumerData = $this->customerData->get($order, Creditcard::METHOD_CODE);
+
+        $orderData = $this->orderDataCollector->collectDataForOrder(
+            $order,
+            Creditcard::PLATFORM_CODE,
+            Creditcard::METHOD_CODE,
+            $this->urlProvider,
+            $this->orderLines,
+            $custumerData,
+            $issuer,
+            $recurringType,
+            $vaultToken
+        );
+
+        $storeId = (int)$order->getStoreId();
+        $client = $this->gingerClient->get($storeId);
+        $transaction = $client->createOrder($orderData);
+
+        $this->writeToFile('transaction', $vaultToken.json_encode($transaction));
+
+        return $transaction ?? false;
+    }
+
+    public function mainRecurring()
+    {
+        $recurringOrders = $this->orders->getOrderRecurringCollection();
+
+        foreach ($recurringOrders as $order)
+        {
+            $transaction = $this->prepareGingerOrder($order);
+            if ($transaction)
```

Comment:
use 'early return' to avoid paddings, I've refactored it, but it still seems difficult to understanding, probably you will find way to simplify it
<img width="997" alt="image" src="https://user-images.githubusercontent.com/80386578/159478699-0542e335-df4a-428b-9135-c90d3ad79779.png">

Reply:
probably after removing TODO comments It will look better

Reply:
Unfortunately, the early return principle is not appropriate here. This function process all recurring orders which should be paid. With early return this function will stop working and some orders stay unprocessed. 
I will try  `continue` instead `return`